source/opt/unify_const_pass.cpp

```diff
@@ -0,0 +1,184 @@
+  // result id and returns that result id. If an existing result id is found,
+  // returns the existing result id.
+  uint32_t LookupEquivalentResultFor(const ir::Instruction& inst) {
+    auto keys = GetLookUpKeys(inst);
+    auto* node = root_.get();
+    for (uint32_t key : keys) {
+      node = node->GetOrCreateTrieNodeFor(key);
+    }
+    if (node->result_id() == 0) {
+      node->SetResultId(inst.result_id());
+    }
+    return node->result_id();
+  }
+
+ private:
+  // The trie node to store result ids.
+  class Node {
+   public:
+    using TrieNodeMap =
+        std::unordered_map<uint32_t, std::unique_ptr<Node>>;
+
+    Node() : result_id_(0), next_() {}
+    uint32_t result_id() const { return result_id_; }
+
+    // Sets the result id stored in this node.
+    void SetResultId(uint32_t id) { result_id_ = id; }
+
+    // Searches for the child trie node with the given key. If the node is
+    // found, returns that node. Otherwise creates an empty child node with
+    // that key and returns that newly created node.
+    Node* GetOrCreateTrieNodeFor(uint32_t key) {
+      auto iter = next_.find(key);
+      if (iter == next_.end()) {
+        // insert a new node and return the node.
+        return next_.insert(std::make_pair(key, MakeUnique<Node>()))
+            .first->second.get();
+      }
+      return iter->second.get();
+    }
+
+   private:
+    // The result id stored in this node. 0 means this node is empty.
+    uint32_t result_id_;
+    // The mapping from the keys to the child nodes of this node.
+    TrieNodeMap next_;
+  };
+
+  // Extracts the opcode, result type id and operand words from an instruction.
+  // Returns a vector of those values in 32-bit unsigned integers.
+  std::vector<uint32_t> GetLookUpKeys(const ir::Instruction& inst) {
+    std::vector<uint32_t> keys;
+    // Need to use the opcode, otherwise there might be a conflict with the
+    // following case when <op>'s binary value equals xx's id:
+    //  OpSpecConstantOp tt <op> yy zz
+    //  OpSpecConstantComposite tt xx yy zz;
+    keys.push_back(static_cast<uint32_t>(inst.opcode()));
+    for (uint32_t i = 0; i < inst.NumOperands(); i++) {
+      if (inst.GetOperand(i).type == SPV_OPERAND_TYPE_RESULT_ID) continue;
+      keys.insert(keys.end(), inst.GetOperand(i).words.cbegin(),
+                  inst.GetOperand(i).words.cend());
+    }
+    return keys;
+  }
+
+  std::unique_ptr<Node> root_;  // The root node of the trie.
+};
+}  // anonymous namespace
+
+bool UnifyConstantPass::Process(ir::Module* module) {
+  bool modified = false;
+  ResultIdTrie defined_constants;
+  analysis::DefUseManager def_use_mgr(module);
+
+  for (ir::Instruction& inst : module->types_values()) {
+    // Do not handle the instruction when there are decorations upon the result
+    // id.
+    if (def_use_mgr.GetAnnotations(inst.result_id()).size() != 0) {
+      continue;
+    }
+
+    // The overall algorithm is to store the result ids of all the eligible
```

Comment:
Great comment.  Describes all the subtle points, including some I had not anticipated.